gemd/ingest/impl.py

```diff
@@ -0,0 +1,208 @@
+
+    Parameters
+    ----------
+    material: MaterialRun
+        The `material` for the returned MeasurementRun
+    name: str
+        The name of the measurement.  Defaults to
+        `template.name` if `template` is defined.
+    template: MeasurementTemplate
+        The MeasurementTemplate.
+
+    Returns
+    --------
+    MeasurementRun
+        A MeasurementRun with linked material, spec and template
+
+    """
+    if name is None:
+        if template is None:
+            raise ValueError("Either a name or a template must be provided")
+        else:
+            name = template.name
+    my_measurement_spec = MeasurementSpec(name, template=template)
+    my_measurement_run = MeasurementRun(name, spec=my_measurement_spec, material=material)
+
+    return my_measurement_run
+
+
+def add_attribute(target: Union[HasConditions, HasParameters, HasProperties],
+                  template: AttributeTemplate,
+                  value: Union[BaseValue, str, float, int]):
+    """
+    Add an attribute to a GEMD object.
+
+    Parameters
+    ----------
+    target: BaseObejct
+        The object to attach the attribute to
+    template: AttributeTemplate
+        The AttributeTemplate for the attribute.
+    value: Union[BaseValue, str, float, int])
+        The value for the attribute.  Accepts any GEMD Value type, or will
+        attempt to generate an appropriate Value given a str, float or int.
+
+    Returns
+    --------
+    BaseAttribute
+        The generated attribute
+
+    """
+    attr_class = Property if isinstance(template, PropertyTemplate) else \
+        Condition if isinstance(template, ConditionTemplate) else \
+        Parameter if isinstance(template, ParameterTemplate) else \
+        None
+    if not isinstance(value, BaseValue):
+        if isinstance(template.bounds, RealBounds):
+            value = NominalReal(value, units=template.bounds.default_units)
+        elif isinstance(template.bounds, IntegerBounds):
+            value = NominalInteger(value)
+        elif isinstance(template.bounds, CategoricalBounds):
+            value = NominalCategorical(value)
+        elif isinstance(template.bounds, CompositionBounds):
+            value = EmpiricalFormula(value)
+        elif isinstance(template.bounds, MolecularStructureBounds):
+            value = Smiles(value)
+        else:
+            raise ValueError(f"Unrecognized bound type in template {type(template.bounds)}")
+
+    attribute = attr_class(name=template.name, value=value, template=template)
+
+    if isinstance(target, MaterialSpec):
+        if attr_class is Property:
+            target.properties.append(PropertyAndConditions(property=attribute))
+        elif attr_class is Condition:
+            if len(target.properties) == 0:
+                raise ValueError("Cannot add a condition to a MaterialSpec "
+                                 "before it has at least one property.")
+            target.properties[-1].conditions.append(attribute)
```

Comment:
I don't like the implicit logic here (you must add the property and then add each conditions, then add the second property and then add each of its conditions, etc.). But I don't see a clearly better way to do it, given the nature of `PropertyAndConditions`.